apache/incubator-nuttx
# SAMA5D2x SDMMC peripheral support (SD Card driver)

## Summary

- SDMMC peripheral support for [SAMA5D2](http://ww1.microchip.com/downloads/en/devicedoc/ds60001476b.pdf) series microprocessors from Microchip
- provides an MMC/SD Card driver that works for reading and writing
- can be used with or without DMA
- UHS_SDR50, UHS_DDR50, and UHS_SDR104 high-speed modes are supported, with  
  read/write speeds of 400Mbit/s in UHS_DDR50 (50MHz Double Data Rate) mode.
- This is a port of the NuttX imxrt_usdhc.c driver, as the SDMMC peripheral
  uses a similar Cadence IP block.
- Basic instructions are in the README.txt.

## Impact

- None if you don't enable the SDMMC peripheral using the Kconfig system.
- Enables reading and writing to MMC or SD Cards. Only SD Cards have been tested.
- VFAT filesystem has been tested and works. On the SAMA5D2-XULT board, reading and writing to a bootable SD Card is supported.

## Testing

- Only tested on the SAMA5D2-XULT board from Microchip.
- Only manual read and write testing on a VFAT-formatted memory card has been done.
- Driver was developed with extensive use of a logic analyzer to ensure the SDMMC driver was functioning correctly.
- Only UHS_DDR50 mode was tested, as all my SD Cards support only that.

## How to Test

1. Format an SD Card with VFAT format. 
1. Put some text files on the card.
1. Insert the card into the SAMA5D2-XULT SD Card slot.
1. Build the sama5d2-xult:sdmmcnsh configuration
1. Load NuttX onto the board
1. /mnt/mmcsd1 should have your files. You can cd, ls, cp, rm, and cat
   to test the filesystem.

## TODO / Limitations

- e.MMC should work, but was not tested
- Performance tests have not been done
- UHS_SDR50 and UHS_SDR104 modes were not tested as I don't have any
  SD Cards that support those modes.

## Notes

- nxstyle shows some formatting errors, but I couldn't find good ways to make them go away without mangling the files. I welcome ideas on how to better format the code so nxstyle passes!
